### ISSUE TITLE
Lazy evaluation in TTNN

### ISSUE THREAD

**Issue by philei-tt** [tenstorrent/tt-metal#31759]
### Is your feature request related to a problem? Please describe

We lack a lot of opportunities to optimize execution in TTNN, since all ops are eager. Since device operations are stateful (invoke returns attributes + tensor arguments, and then the tt-nn backend knows how to dispatch that to the device), it should be possible to delay execution till we actually need to use the results.
This is very similar to how [mlx](https://github.com/ml-explore/mlx) works, and compilation feature of [torch 2.0](https://docs.pytorch.org/tutorials/intermediate/torch_compile_tutorial.html), [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) or [mlx](https://ml-explore.github.io/mlx/build/html/usage/compile.html)

This would allow us to do graph optimizations like: 
- dead code elimination
- op fusion
- Constant folding
- CSE
- Canonicalization
- Layout propagation & normalization
- Automatic deallocation
- Recomputation/rematerialization (if low on memory)

And do some other optimizations:
- Better kernel /sharding selection
- Automatic metal trace
- Opens possibility for autograd in tt-nn 

In addition, this should simplify implementation of tools like graph capture (should help with hangs in no dispatch, since in lazy we can allow some ops to do partial dispatch), pre-compilation of kernels.
Also, lazy execution gives control over how to execute the graph (e.g., we can use current tt-nn as backend, but we could also use MLIR to generate a backend for he op)

### Describe the solution you'd like

- We should move Tensor into metal (https://github.com/tenstorrent/tt-metal/issues/30314) and then make ttnn::Tensor a lazy Tensor. Lazy Tensor means that it only stores meta information (TensorSpec, device, StorageType) and connections (the operation used to produce it and the input lazy tensors). When it's materialized, it now also stores tt_meal::Tensor. This idea is similar to how mlx works -- we don't have any specific graph and node structures. Tensors are nodes in the execution graph, and lazy and eager modes don't require any code modification from the user. 
- We need to wrap current device operations in a deferred "LazyDeviceOperation." Ideally, this should require zero code changes to current device operations. 
- We should update execution logic, i.e., store operation and return non-materialized Tensor if we're in lazy mode, otherwise follow the current eager path. 
- We need a few graph utilities (e.g., topological sort) to implement materialization
- We should add and test passes for graph optimization, like fusion pass, to verify that it's possible to optimize the graph
- I think we should try to rebuild the graph capture with a lazy graph.

### Describe alternatives you've considered

PoC of lazy execution: https://github.com/tenstorrent/tt-metal/pull/31492 + simple unary op fusion pass. (Note, this branch is going to grow and get new features)
